benchmarks/PNP/pnp_adaptive.cpp

```diff
@@ -0,0 +1,548 @@
+      copy_dvector_to_vector_function(&solu_fasp, &solutionUpdate, &cation_dofs, &cation_dofs);
+      copy_dvector_to_vector_function(&solu_fasp, &solutionUpdate, &anion_dofs, &anion_dofs);
+      copy_dvector_to_vector_function(&solu_fasp, &solutionUpdate, &potential_dofs, &potential_dofs);
+
+      // update solution and reset solutionUpdate
+      printf("\tupdate solution...\n"); fflush(stdout);
+      relative_residual = update_solution_pnp(
+        &cationSolution,
+        &anionSolution,
+        &potentialSolution,
+        &(solutionUpdate[0]),
+        &(solutionUpdate[1]),
+        &(solutionUpdate[2]),
+        relative_residual,
+        initial_residual,
+        &L_pnp,
+        &bc,
+        &newtparam
+      );
+      if (relative_residual < 0.0) {
+        printf("Newton backtracking failed!\n");
+        printf("\tresidual has not decreased after damping %d times\n", newtparam.damp_it);
+        printf("\tthe relative residual is %e\n", relative_residual);
+        relative_residual *= -1.0;
+      }
+
+      // compute residual
+      L_pnp.CatCat = cationSolution;
+      L_pnp.AnAn = anionSolution;
+      L_pnp.EsEs = potentialSolution;
+      assemble(b_pnp, L_pnp);
+      bc.apply(b_pnp);
+      relative_residual = b_pnp.norm("l2") / initial_residual;
+
+      if (newton_iteration == 1)
+        printf("\trelative nonlinear residual after 1 iteration has l2-norm of %e\n", relative_residual);
+      else
+        printf("\trelative nonlinear residual after %d iterations has l2-norm of %e\n", newton_iteration, relative_residual);
+
+      // write computed solution to file
+      printf("\tsolved linear system successfully!\n"); fflush(stdout);
+      cationFile << cationSolution;
+      anionFile << anionSolution;
+      potentialFile << potentialSolution;
+
+      // compute solution error
+      printf("\nCompute the error\n"); fflush(stdout);
+      dolfin::Function Error1(analyticCation);
+      dolfin::Function Error2(analyticAnion);
+      dolfin::Function Error3(analyticPotential);
+      *(Error1.vector()) -= *(cationSolution.vector());
+      *(Error2.vector()) -= *(anionSolution.vector());
+      *(Error3.vector()) -= *(potentialSolution.vector());
+      double cationError = 0.0;
+      double anionError = 0.0;
+      double potentialError = 0.0;
+      L2Error::Form_M L2error1(mesh,Error1);
+      cationError = assemble(L2error1);
+      L2Error::Form_M L2error2(mesh,Error2);
+      anionError = assemble(L2error2);
+      L2Error::Form_M L2error3(mesh,Error3);
+      potentialError = assemble(L2error3);
+      printf("\tcation l2 error is:     %e\n", cationError);
+      printf("\tanion l2 error is:      %e\n", anionError);
+      printf("\tpotential l2 error is:  %e\n", potentialError);
+
+      // print error
+      // cationFile << cationSolution;
+      // anionFile << anionSolution;
+      // potentialFile << potentialSolution;
+    }
+
+    if (relative_residual < nonlinear_tol)
```

Comment:
So the mesh adaptivity has to be outside the newton loop ? What happened if you put the mesh adaptivity at each newton step ?